src/vivarium/framework/resource.py

```diff
@@ -0,0 +1,123 @@
+"""
+=====================
+Dependency Management
+=====================
+
+This module provides a tool to manage dependencies on resources within a
+``Vivarium`` simulation. These resources take the form of things that can be
+created and utilized by components, for example columns in the
+:mod:`state table <vivarium.framework.population>`
+or :mod:`named value pipelines <vivarium.framework.values>`. Because these
+need to be created before they can be used, they are sensitive to ordering.
+The intent behind this tool is to provide an interface that allows other
+managers to register resources with the dependency manager and in turn ask for
+ordered sequences of these resources according to their dependencies or raise
+exceptions if this is not possible.
+
+"""
+
+import networkx as nx
+
+from vivarium.exceptions import VivariumError
+
+
+class ResourceError(VivariumError):
+    """Error raised when a dependency requirement is violated"""
+    pass
+
+
+RESOURCE_TYPES = {'value', 'value_source', 'value_modifier', 'column'}
+
+
+class ResourceProducer:
+
+    def __init__(self, resource_type, resource_names, producer, dependencies):
```

Comment:
some type hints would be nice here. Woulda helped me reason about how this is expected to be used